packages/server/graphql/mutations/endCheckIn.ts

```diff
@@ -0,0 +1,303 @@
+      .getAll(teamId, {index: 'teamId'})
+      .filter({isActive: true})
+      .run()
+  ])
+
+  const activeAgendaItemIds = activeAgendaItems.map(({id}) => id)
+  const userIds = meetingMembers.map(({userId}) => userId)
+  const meetingPhase = getMeetingPhase(phases)
+  const pinnedAgendaItems = await getPinnedAgendaItems(teamId)
+  const isKill = meetingPhase && ![AGENDA_ITEMS, LAST_CALL].includes(meetingPhase.phaseType)
+  await Promise.all([
+    isKill ? undefined : archiveTasksForDB(doneTasks, meetingId),
+    isKill ? undefined : clearAgendaItems(teamId),
+    isKill ? undefined : clonePinnedAgendaItems(pinnedAgendaItems),
+    updateTaskSortOrders(userIds, tasks),
+    r
+      .table('NewMeeting')
+      .get(meetingId)
+      .update(
+        {
+          agendaItemCount: activeAgendaItems.length,
+          commentCount: (r
+            .table('Comment')
+            .getAll(r.args(activeAgendaItemIds), {index: 'threadId'})
+            .count()
+            .default(0) as unknown) as number,
+          taskCount: tasks.length
+        },
+        {nonAtomic: true}
+      )
+      .run()
+  ])
+
+  return {updatedTaskIds: [...tasks, ...doneTasks].map(({id}) => id)}
+}
+
+export default {
+  type: new GraphQLNonNull(EndCheckInPayload),
+  description: 'Finish a check-in meeting',
+  args: {
+    meetingId: {
+      type: new GraphQLNonNull(GraphQLID),
+      description: 'The meeting to end'
+    }
+  },
+  async resolve(_source, {meetingId}, context: GQLContext) {
+    const {authToken, socketId: mutatorId, dataLoader} = context
+    const r = await getRethink()
+    const operationId = dataLoader.share()
+    const subOptions = {mutatorId, operationId}
+    const now = new Date()
+    const viewerId = getUserId(authToken)
+
+    // AUTH
+    const meeting = (await r
+      .table('NewMeeting')
+      .get(meetingId)
+      .default(null)
+      .run()) as MeetingAction | null
+    if (!meeting) return standardError(new Error('Meeting not found'), {userId: viewerId})
+    const {endedAt, facilitatorStageId, meetingNumber, phases, teamId, meetingType} = meeting
+
+    // VALIDATION
+    if (!isTeamMember(authToken, teamId) && authToken.rol !== 'su') {
+      return standardError(new Error('Team not found'), {userId: viewerId})
+    }
+    if (endedAt) return standardError(new Error('Meeting already ended'), {userId: viewerId})
+
+    // RESOLUTION
+    const currentStageRes = findStageById(phases, facilitatorStageId)
+    if (!currentStageRes) {
+      return standardError(new Error('Cannot find facilitator stage'), {userId: viewerId})
+    }
+    const {stage} = currentStageRes
+    const phase = getMeetingPhase(phases)
+    stage.isComplete = true
+    stage.endAt = now
+
+    const completedCheckIn = ((await r
+      .table('NewMeeting')
+      .get(meetingId)
+      .update(
+        {
+          endedAt: now,
+          phases
+        },
+        {returnChanges: true}
+      )('changes')(0)('new_val')
```

Comment:
it's possible (not probable, though) that this gets called 2x at the exact same time.
in that case, we'll wait to handle the case that an update did not result in a `new_val`.
easiest thing to do is add a `.default(null)` and then see `if (!completedCheckIn)...`

Reply:
Interesting - it could get called twice because of something going wrong in the codebase or would the user be able to trigger it twice?

Reply:
user could trigger it twice. a good pen-tester will always rapid fire the same query over & over hoping to break something. few years back, a guy figured out he could get free starbucks giftcards doing this. albeit ours isn't as exciting as that...

Reply:
haha good to know!